### PR TITLE
Link time optimization

### PR DESCRIPTION
**Pull request by igudich**
1. LTO is disabled for tests as far as it is not important.
2. CMAKE_INTERPROCEDURAL_OPTIMIZATION=OFF can be used to disable LTO for libs.
3. -DDYNAMIC_LIB forces the shared libraries building, that also decreases link time.

For dev purposes   `-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=OFF -DDYNAMIC_LIB=ON` is recommended.

### CODE REVIEW

File: core/dataset.cpp [hunk 1]
```diff
@@ -249,6 +249,19 @@ bool checkCorrespondingDenseCoords(const Dataset &dataset,
...
+  const auto dsLItems{dsLabels.items()};
+  for (const auto & [ nm, v ] : otLabels) {
+    if (auto iter = dsLItems.find(nm); iter == dsLItems.end()) {
+      return false;
+    } else {
+      if (*iter->second.first != v) {
+        return false;
+      }
+    }
+  }
```
Comment (SimonHeybrock) [inline core/dataset.cpp]: You have some things that were part of another unrelated PR here.

File: core/dataset.cpp [hunk 2]
```diff
@@ -279,6 +292,12 @@ void Dataset::setData(const std::string &name, const DataConstProxy &data) {
   if (dim != Dim::Invalid) {
     setSparseCoord(name, data.coords()[dim]);
   }
+
+  for (const auto & [ nm, v ] : data.labels()) {
+    if (v.dims().sparse()) {
+      setSparseLabels(name, name, v);
+    }
+  }
```
Comment (SimonHeybrock) [inline core/dataset.cpp]: unrelated

File: core/include/scipp/core/variable.h [hunk 3]
```diff
@@ -53,7 +53,7 @@ using VariableConceptHandle = VariableConceptHandle_impl<
 ///
 /// This is the most generic representation for a multi-dimensional array of
 /// data. More operations are supportd by the partially-typed VariableConceptT.
-class VariableConcept {
+class SCIPP_CORE_EXPORT VariableConcept {
```
Comment (SimonHeybrock) [inline core/include/scipp/core/variable.h]: Is this required? I think `VariableConcept` should only be an internal and is not required to be visible to other libraries.

**Comment by SimonHeybrock** [comment]
>     2\. -DDISABLE_LTO disables LTO for libraries for faster linking.
> 
>     3\. -DDYNAMIC_LIB forces the shared libraries building, that also decreases link time.
> 
> 
> For dev purposes `-DDISABLE_LTO=ON -DDYNAMIC_LIB=ON` is recommended.

Can you put this in the readme somewhere, otherwise it will be lost and forgotten.

File: core/include/scipp/core/variable.h (hunk 3 above)
Comment (igudich) [reply core/include/scipp/core/variable.h]: It doesn't link without this...

File: CMakeLists.txt [hunk 4]
```diff
@@ -58,6 +58,8 @@ else()
 endif()
 
 option(WITH_CTEST "Enable ctest integration of tests" ON)
+option(DISABLE_LTO "Disable LTO for faster linking" OFF)
```
Comment (DanNixon) [inline CMakeLists.txt]: Could this be changed to `ENABLE_LTO`?

Ideally setting a boolean to `ON` would enable something, not disable it.

File: neutron/CMakeLists.txt [hunk 5]
```diff
@@ -1,19 +1,36 @@
...
+         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+         ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(${TARGET_NAME} SYSTEM
                            PUBLIC ${CMAKE_BINARY_DIR}/Eigen-src
                            PRIVATE "../range-v3/include")
-set_target_properties(scipp-neutron PROPERTIES POSITION_INDEPENDENT_CODE TRUE)
+
+if(NOT ENABLE_LTO)
+    set_property(TARGET ${TARGET_NAME} PROPERTY INTERPROCEDURAL_OPTIMIZATION FALSE)
+endif(NOT ENABLE_LTO)
```
Comment (SimonHeybrock) [inline neutron/CMakeLists.txt]: Instead of doing this in every module, change it globally here: https://github.com/scipp/scipp/blob/master/CMakeLists.txt#L57

File: CMakeLists.txt [hunk 6]
```diff
@@ -58,6 +58,8 @@ else()
 endif()
 
 option(WITH_CTEST "Enable ctest integration of tests" ON)
+option(ENABLE_LTO "Disable LTO for faster linking" ON)
```
Comment (SimonHeybrock) [inline CMakeLists.txt]: Maybe this flag is completely unnecessary? After all we can just set `CMAKE_INTERPROCEDURAL_OPTIMIZATION` (use default TRUE, instead of forcing TRUE in line 57?)?

File: neutron/CMakeLists.txt (hunk 5 above)
Comment (igudich) [reply neutron/CMakeLists.txt]: I want the LTO disabled for tests forever and a flag for libraries. So either you would do this manually for tests or for libs, right?

Comment (SimonHeybrock) [reply neutron/CMakeLists.txt]: Yes, keep it manually disabled for tests, everything else using the global cmake flag.

Comment (igudich) [reply neutron/CMakeLists.txt]: I can remove this flag and we can use CMAKE_INTERPROCEDURAL_OPTIMIZATION wich also affects producing the fat lto objects, but all the optimization would be ignored with -flto switched off.

Comment (SimonHeybrock) [reply neutron/CMakeLists.txt]: Do we care about fat lto objects?

File: CMakeLists.txt (hunk 6 above)
Comment (igudich) [reply CMakeLists.txt]: How would you switch it off using command line?

Comment (SimonHeybrock) [reply CMakeLists.txt]:
```
if(NOT CMAKE_INTERPROCEDURAL_OPTIMIZATION)
    set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
endif()
```

or similar, see also https://github.com/scipp/scipp/blob/b3dafad98a57c79f66d4e69697aae697045b0918/CMakeLists.txt#L16-L23.
